ide-plugin/src/org/jetbrains/plugins/ruby/ruby/codeInsight/types/RubyReturnTypeProvider.kt

```diff
@@ -37,6 +48,36 @@ class RubyReturnTypeProvider : AbstractRubyTypeProvider() {
             }
             return originType
         }
+        if (expr is RIdentifier && expr.isParameter) {
+            val method = expr.parentContainer as? RMethod ?: return null
+            val moduleName = (method.parentContainer as? RElementWithFQN)?.fqn?.fullPath ?: return null
+            val indexOfArgument = method.arguments.indexOfFirst { it.identifier == expr }
+
+            val info = MethodInfo(ClassInfo(moduleName), method.fqn.shortName, RVisibility.PUBLIC)
+
+            val callInfos: List<CallInfo> = RSignatureProviderImpl().getRegisteredCallInfos(info)
+
+            return callInfos.map {
+                if (indexOfArgument in 0 until it.argumentsTypes.size) {
+                    RTypeFactory.createTypeClassName(it.argumentsTypes[indexOfArgument], expr)
+                } else {
+                    REmptyType.INSTANCE
+                }
+            }.filter { it != REmptyType.INSTANCE }.distinctBy { it.name }.unionTypes()
+
+        }
         return null
     }
+}
+
+private fun List<RType>.unionTypes(startIndex: Int = 0, endIndex: Int = size - 1): RType {
+    if (endIndex < startIndex) {
+        return REmptyType.INSTANCE
+    }
+
+    if (startIndex == endIndex) {
+        return this[startIndex]
+    }
+    val m = (startIndex + endIndex) / 2
+    return RTypeUtil.union(this.unionTypes(startIndex, m), this.unionTypes(m + 1, endIndex))
```

Comment:
I think using `java.util.List#subList` will be more traditional approach.

Reply:
Don't really understand why you think that `java.util.List#subList` is better in this case. My variant requires less memory and doesn't have any cons compared to `java.util.List#subList`

Reply:
Your variant is re-implementation of the same thing, which does not make any sense. The memory consumption difference is negligible at the same time.

Just in case: when using `subList` you don't need any parameters to the function which makes it safer to use. (now it can be used in such a way to throw an AIOOBE)